File: app/src/main/java/uz/efir/muazzin/Preferences.java
package uz.efir.muazzin;

import android.content.Context;
import android.content.SharedPreferences;
import android.location.Location;

import androidx.annotation.NonNull;

/**
 * Helper class to store and retrieve user settings to/from shared preferences
 */
public class Preferences {

    private static final String PREFERENCE_FILENAME_SUFFIX = "_preferences";
    private static final String KEY_BASMALA = "key_bismillah_on_boot_up";
    private static final String KEY_CALCULATION_METHOD_INDEX = "calculation_method_index";
    private static final String KEY_NOTIFICATION_METHOD = "notification_method_";
    private static final String KEY_LATITUDE = "location_latitude";
    private static final String KEY_LONGITUDE = "location_longitude";
    private static final String KEY_OFFSET_MINUTES = "offset_minutes";
    private static Preferences sPreferences;
    private final SharedPreferences mSharedPreferences;

    private Preferences(@NonNull Context context) {
        mSharedPreferences = context.getSharedPreferences(context.getPackageName().concat(PREFERENCE_FILENAME_SUFFIX), Context.MODE_PRIVATE);
    }

    public static Preferences getInstance(Context context) {
        return sPreferences == null ? (sPreferences = new Preferences(context)) : sPreferences;
    }

    public boolean dontNotifySunrise() {
        return CONSTANT.NOTIFICATION_NONE == getNotificationMethod(CONSTANT.SUNRISE);
    }

    public boolean getBasmalaEnabled() {
        return mSharedPreferences.getBoolean(KEY_BASMALA, false);
    }

    public void setBasmalaEnabled(boolean enabled) {
        final SharedPreferences.Editor editor = mSharedPreferences.edit();
        editor.putBoolean(KEY_BASMALA, enabled);
        editor.apply();
    }

    public int getNotificationMethod(int time) {
        return mSharedPreferences.getInt(KEY_NOTIFICATION_METHOD.concat(Integer.toString(time)), CONSTANT.SUNRISE == time ? CONSTANT.NOTIFICATION_NONE : CONSTANT.NOTIFICATION_DEFAULT);
    }

    public void setNotificationMethod(short time, int method) {
        final SharedPreferences.Editor editor = mSharedPreferences.edit();
        editor.putInt(KEY_NOTIFICATION_METHOD.concat(Short.toString(time)), method);
        editor.apply();
    }

    public int getCalculationMethodIndex() {
        return mSharedPreferences.getInt(KEY_CALCULATION_METHOD_INDEX, CONSTANT.DEFAULT_CALCULATION_METHOD);
    }

    public void setCalculationMethodIndex(int index) {
        final SharedPreferences.Editor editor = mSharedPreferences.edit();
        editor.putInt(KEY_CALCULATION_METHOD_INDEX, index);
        editor.apply();
    }

    public Location getLocation() {
        Location location = new Location("");
        // Kaaba coordinates: 21.4225° N, 39.8261° E
        location.setLatitude(mSharedPreferences.getFloat(KEY_LATITUDE, 21.4225f));
        location.setLongitude(mSharedPreferences.getFloat(KEY_LONGITUDE, 39.8261f));
        return location;
    }

    public void setLocation(@NonNull Location location) {
        final SharedPreferences.Editor editor = mSharedPreferences.edit();
        editor.putFloat(KEY_LATITUDE, (float) location.getLatitude());
        editor.putFloat(KEY_LONGITUDE, (float) location.getLongitude());
        editor.apply();
    }

    public boolean isLocationSet() {
        return mSharedPreferences.contains(KEY_LATITUDE) && mSharedPreferences.contains(KEY_LONGITUDE);
    }

    public boolean isCalculationMethodSet() {
        return mSharedPreferences.contains(KEY_CALCULATION_METHOD_INDEX);
    }

    public int getOffsetMinutes() {
        return mSharedPreferences.getInt(KEY_OFFSET_MINUTES, 0);
    }

    public void setOffsetMinutes(int minutes) {
        final SharedPreferences.Editor editor = mSharedPreferences.edit();
        editor.putInt(KEY_OFFSET_MINUTES, minutes);
        editor.apply();
    }
}

File: app/src/main/kotlin/uz/efir/muazzin/Preferences.kt
package uz.efir.muazzin

import android.content.Context
import android.content.SharedPreferences
import android.location.Location
import android.util.Log
import androidx.core.content.edit

class Preferences private constructor(context: Context) {
    private val mSharedPreferences: SharedPreferences = context.getSharedPreferences(
        context.packageName + PREFERENCE_FILENAME_SUFFIX, Context.MODE_PRIVATE
    )

    init {
        migrate()
    }

    /**
     * Migrate persisted preferences forward when their schema changes.
     * Runs once per upgrade, gated by KEY_PREFS_VERSION.
     */
    private fun migrate() {
        val current = mSharedPreferences.getInt(KEY_PREFS_VERSION, 0)
        if (current >= PREFS_VERSION) return

        Log.i("Preferences", "Migrating preferences from v$current to v$PREFS_VERSION")

        @Suppress("KotlinConstantConditions")
        if (current < 1 && isCalculationMethodSet) {
            // v1: collapse Karachi Shafi/Hanafi duplicate. Old order:
            //   0=EGYPTIAN, 1=KARACHI Shafi, 2=KARACHI Hanafi, 3=N. AMERICA, 4=MWL, 5=UMM_AL_QURA
            // New order:
            //   0=EGYPTIAN, 1=KARACHI, 2=N. AMERICA, 3=MWL, 4=UMM_AL_QURA
            // Old indices 0 and 1 stay; old indices >= 2 shift down by 1.
            val old = mSharedPreferences.getInt(KEY_CALCULATION_METHOD_INDEX, -1)
            if (old >= 2) {
                calculationMethodIndex = old - 1
            }
        }

        mSharedPreferences.edit { putInt(KEY_PREFS_VERSION, PREFS_VERSION) }
    }

    fun getNotificationMethod(time: Int): Int {
        val default =
            if (time == PrayerTime.SUNRISE.ordinal) NotificationMethod.NONE else NotificationMethod.DEFAULT
        return mSharedPreferences.getInt(KEY_NOTIFICATION_METHOD + time.toString(), default.ordinal)
    }

    fun setNotificationMethod(time: Int, method: Int) {
        mSharedPreferences.edit {
            putInt(KEY_NOTIFICATION_METHOD + time.toString(), method)
        }
    }

    var calculationMethodIndex: Int
        get() = mSharedPreferences.getInt(
            KEY_CALCULATION_METHOD_INDEX, CalculationMethodCatalog.DEFAULT_INDEX
        )
        set(index) {
            mSharedPreferences.edit {
                putInt(
                    KEY_CALCULATION_METHOD_INDEX, index
                )
            }
        }

    var location: Location
        get() {
            val location = Location("")
            // Kaaba coordinates: 21.4225° N, 39.8261° E
            location.latitude = mSharedPreferences.getFloat(
                KEY_LATITUDE, 21.4225f
            ).toDouble()
            location.longitude = mSharedPreferences.getFloat(
                KEY_LONGITUDE, 39.8261f
            ).toDouble()
            return location
        }
        set(location) {
            mSharedPreferences.edit {
                putFloat(
                    KEY_LATITUDE, location.latitude.toFloat()
                )
                putFloat(
                    KEY_LONGITUDE, location.longitude.toFloat()
                )
            }
        }

    val isLocationSet: Boolean
        get() = mSharedPreferences.contains(KEY_LATITUDE) &&
                mSharedPreferences.contains(KEY_LONGITUDE)

    val isCalculationMethodSet: Boolean
        get() = mSharedPreferences.contains(KEY_CALCULATION_METHOD_INDEX)

    var offsetMinutes: Int
        get() = mSharedPreferences.getInt(KEY_OFFSET_MINUTES, 0)
        set(minutes) {
            mSharedPreferences.edit {
                putInt(KEY_OFFSET_MINUTES, minutes)
            }
        }

    companion object {
        private const val PREFERENCE_FILENAME_SUFFIX = "_preferences"
        private const val KEY_CALCULATION_METHOD_INDEX = "calculation_method_index"
        private const val KEY_NOTIFICATION_METHOD = "notification_method_"
        private const val KEY_LATITUDE = "location_latitude"
        private const val KEY_LONGITUDE = "location_longitude"
        private const val KEY_OFFSET_MINUTES = "offset_minutes"
        private const val KEY_PREFS_VERSION = "prefs_version"

        // Bump and add a migrate() branch when the persistence schema changes.
        private const val PREFS_VERSION = 1

        @Volatile
        private var sPreferences: Preferences? = null
        fun getInstance(context: Context): Preferences {
            return sPreferences ?: synchronized(this) {
                sPreferences ?: Preferences(context.applicationContext).also { sPreferences = it }
            }
        }
    }
}
